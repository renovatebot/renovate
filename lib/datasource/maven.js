const got = require('got');
const url = require('url');
const fs = require('fs-extra');
const { XmlDocument } = require('xmldoc');

module.exports = {
  getPkgReleases,
};

// eslint-disable-next-line no-unused-vars
async function getPkgReleases(purl, config) {
  const versions = [];
  const dependency = getDependencyParts(purl);
  const repositories = getRepositories(purl);
  if (repositories.length < 1) {
    logger.error(`No repositories defined for ${dependency.display}`);
    return null;
  }
  logger.debug(
    `Found ${repositories.length} repositories for ${dependency.display}`
  );
  for (let i = 0; i < repositories.length; i += 1) {
    const repoUrl = repositories[i];
    logger.debug(
      `Looking up ${dependency.display} in repository #${i} - ${repoUrl}`
    );
    const mavenMetadata = await downloadMavenMetadata(dependency, repoUrl);
    if (mavenMetadata) {
      const newVersions = extractVersions(mavenMetadata).filter(
        version => !versions.includes(version)
      );
      versions.push(...newVersions);
      logger.debug(`Found ${newVersions.length} new versions for ${dependency.display} in repository ${repoUrl}`); // prettier-ignore
    }
  }

  if (versions.length === 0) {
    logger.warn(`No versions found for ${dependency.display} in ${repositories.length} repositories`); // prettier-ignore
    return null;
  }
  logger.debug(`Found ${versions.length} versions for ${dependency.display}`);

  return {
    ...dependency,
    releases: versions.map(v => ({ version: v })),
  };
}

function getDependencyParts(purl) {
  return {
    display: `${purl.namespace}:${purl.name}`,
    group: purl.namespace,
    name: purl.name,
    version: purl.version,
    dependencyUrl: generateMavenUrl(purl),
  };
}

function getRepositories(purl) {
  if (!purl.qualifiers || !purl.qualifiers.repository_url) {
    return [];
  }
  return purl.qualifiers.repository_url.split(',').map(repoUrl => {
    if (!repoUrl.endsWith('/')) {
      return repoUrl + '/';
    }
    return repoUrl;
  });
}

async function downloadMavenMetadata(dependency, repoUrl) {
  const pkgUrl = new url.URL(
    `${dependency.dependencyUrl}/maven-metadata.xml`,
    repoUrl
  );
  let mavenMetadata;
  switch (pkgUrl.protocol) {
    case 'file:':
      mavenMetadata = await downloadFileProtocol(pkgUrl);
      break;
    case 'http:':
    case 'https:':
      mavenMetadata = await downloadHttpProtocol(pkgUrl);
      break;
    default:
      logger.error(
        `Invalid protocol ${pkgUrl.protocol} in repository ${repoUrl}`
      );
      return null;
  }
  if (!mavenMetadata) {
    logger.debug(`${dependency.display} not found in repository ${repoUrl}`);
  }
  return mavenMetadata;
}

function extractVersions(mavenMetadata) {
  const doc = new XmlDocument(mavenMetadata);
  const versions = doc.descendantWithPath('versioning.versions');
  const elements = versions && versions.childrenNamed('version');
  if (!elements) return [];
  return elements.map(el => el.val);
}

async function downloadFileProtocol(pkgUrl) {
  const pkgPath = pkgUrl.toString().replace('file://', '');
  if (!(await fs.exists(pkgPath))) {
    return null;
  }
  return fs.readFile(pkgPath, 'utf8');
}

async function downloadHttpProtocol(pkgUrl) {
  let raw;
  try {
    raw = await got(pkgUrl);
  } catch (err) {
    if (isNotFoundError(err)) {
      logger.debug(`Url not found ${pkgUrl}`);
    } else if (isTemporalError(err)) {
      logger.warn(`Error requesting ${pkgUrl} Error Code: ${err.statusCode}`);
      if (isMavenCentral(pkgUrl)) {
        throw new Error('registry-failure');
      }
    } else {
      logger.warn(
        `Unknown error requesting ${pkgUrl} Error Code: ${err.statusCode}`
      );
    }
    return null;
  }
  return raw.body;
}

function generateMavenUrl(purl) {
  return purl.namespace.replace(/\./g, '/') + `/${purl.name}`;
}

function isMavenCentral(pkgUrl) {
  return pkgUrl.host === 'central.maven.org';
}

function isTemporalError(err) {
  return (
    err.statusCode === 429 || (err.statusCode > 500 && err.statusCode < 600)
  );
}

function isNotFoundError(err) {
  return err.statusCode === 404;
}
