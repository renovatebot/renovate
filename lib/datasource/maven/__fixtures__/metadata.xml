<?xml version="1.0" encoding="UTF-8"?>
<metadata>
  <groupId>org.example</groupId>
  <artifactId>package</artifactId>
  <versioning>
    <latest>2.0.0</latest>
    <release>2.0.0</release>
    <versions>
      <version>1.0.0</version>
      <version>1.0.1</version>
      <version>1.0.2</version>
      <version>2.0.0</version>
    </versions>
    <lastUpdated>20210101000000</lastUpdated>
  </versioning>
</metadata>
