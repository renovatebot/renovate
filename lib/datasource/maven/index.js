const url = require('url');
const fs = require('fs-extra');
const { XmlDocument } = require('xmldoc');
const is = require('@sindresorhus/is');

const { compare } = require('../../versioning/maven/compare');
const { containsPlaceholder } = require('../../manager/maven/extract');
const { downloadHttpProtocol } = require('./util');

module.exports = {
  getPkgReleases,
};

// eslint-disable-next-line no-unused-vars
async function getPkgReleases({ lookupName, registryUrls }) {
  const versions = [];
  const dependency = getDependencyParts(lookupName);
  if (!is.nonEmptyArray(registryUrls)) {
    logger.error(`No repositories defined for ${dependency.display}`);
    return null;
  }
  const repositories = registryUrls.map(repository =>
    repository.replace(/\/?$/, '/')
  );
  logger.debug(
    `Found ${repositories.length} repositories for ${dependency.display}`
  );
  const repoForVersions = {};
  for (let i = 0; i < repositories.length; i += 1) {
    const repoUrl = repositories[i];
    logger.debug(
      `Looking up ${dependency.display} in repository #${i} - ${repoUrl}`
    );
    const mavenMetadata = await downloadMavenXml(
      dependency,
      repoUrl,
      'maven-metadata.xml'
    );
    if (mavenMetadata) {
      const newVersions = extractVersions(mavenMetadata).filter(
        version => !versions.includes(version)
      );
      const latestVersion = getLatestVersion(newVersions);
      if (latestVersion) {
        repoForVersions[latestVersion] = repoUrl;
      }
      versions.push(...newVersions);
      logger.debug(`Found ${newVersions.length} new versions for ${dependency.display} in repository ${repoUrl}`); // prettier-ignore
    }
  }

  if (versions.length === 0) {
    logger.info(`No versions found for ${dependency.display} in ${repositories.length} repositories`); // prettier-ignore
    return null;
  }
  logger.debug(`Found ${versions.length} versions for ${dependency.display}`);
  const latestVersion = getLatestVersion(versions);
  const repoUrl = repoForVersions[latestVersion];
  const dependencyInfo = await getDependencyInfo(
    dependency,
    repoUrl,
    latestVersion
  );

  return {
    ...dependency,
    ...dependencyInfo,
    releases: versions.map(v => ({ version: v })),
  };
}

function getDependencyParts(lookupName) {
  const [group, name] = lookupName.split(':');
  const dependencyUrl = `${group.replace(/\./g, '/')}/${name}`;
  return {
    display: lookupName,
    group,
    name,
    dependencyUrl,
  };
}

async function downloadMavenXml(dependency, repoUrl, dependencyFilePath) {
  const pkgUrl = new url.URL(
    `${dependency.dependencyUrl}/${dependencyFilePath}`,
    repoUrl
  );

  let rawContent;
  switch (pkgUrl.protocol) {
    case 'file:':
      rawContent = await downloadFileProtocol(pkgUrl);
      break;
    case 'http:':
    case 'https:':
      rawContent = await downloadHttpProtocol(pkgUrl);
      break;
    case 's3:':
      logger.debug('Skipping s3 dependency');
      return null;
    default:
      logger.warn(
        `Invalid protocol ${pkgUrl.protocol} in repository ${repoUrl}`
      );
      return null;
  }

  if (!rawContent) {
    logger.debug(`${dependency.display} not found in repository ${repoUrl}`);
    return null;
  }

  try {
    return new XmlDocument(rawContent);
  } catch (e) {
    logger.debug(`Can not parse ${pkgUrl.href} for ${dependency.display}`);
    return null;
  }
}

function extractVersions(metadata) {
  const versions = metadata.descendantWithPath('versioning.versions');
  const elements = versions && versions.childrenNamed('version');
  if (!elements) return [];
  return elements.map(el => el.val);
}

async function downloadFileProtocol(pkgUrl) {
  const pkgPath = pkgUrl.toString().replace('file://', '');
  if (!(await fs.exists(pkgPath))) {
    return null;
  }
  return fs.readFile(pkgPath, 'utf8');
}

function getLatestVersion(versions) {
  if (versions.length === 0) return null;
  return versions.reduce((latestVersion, version) =>
    compare(version, latestVersion) === 1 ? version : latestVersion
  );
}

async function getDependencyInfo(dependency, repoUrl, version) {
  const result = {};
  const path = `${version}/${dependency.name}-${version}.pom`;

  const pomContent = await downloadMavenXml(dependency, repoUrl, path);
  if (!pomContent) return result;

  const homepage = pomContent.valueWithPath('url');
  if (homepage && !containsPlaceholder(homepage)) {
    result.homepage = homepage;
  }

  const sourceUrl = pomContent.valueWithPath('scm.url');
  if (sourceUrl && !containsPlaceholder(sourceUrl)) {
    result.sourceUrl = sourceUrl;
  }

  return result;
}
