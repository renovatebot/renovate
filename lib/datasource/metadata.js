const is = require('@sindresorhus/is');

module.exports = {
  addMetaData,
};

// Use this object to define changelog URLs for packages
// Only necessary when the changelog data cannot be found in the package's source repository
const manualChangelogUrls = {
  npm: {
    'babel-preset-react-app':
      'https://github.com/facebook/create-react-app/releases',
    firebase: 'https://firebase.google.com/support/release-notes/js',
    'flow-bin': 'https://github.com/facebook/flow/blob/master/Changelog.md',
    'react-native':
      'https://github.com/react-native-community/react-native-releases/blob/master/CHANGELOG.md',
  },
  pypi: {
    'pytest-django':
      'https://pytest-django.readthedocs.io/en/latest/changelog.html#changelog',
    django: 'https://github.com/django/django/tree/master/docs/releases',
    djangorestframework:
      'https://www.django-rest-framework.org/community/release-notes/',
    flake8: 'http://flake8.pycqa.org/en/latest/release-notes/index.html',
    'django-storages':
      'https://github.com/jschneier/django-storages/blob/master/CHANGELOG.rst',
    phonenumbers:
      'https://github.com/daviddrysdale/python-phonenumbers/blob/dev/python/HISTORY.md',
    'psycopg2-binary': 'http://initd.org/psycopg/articles/tag/release/',
    'django-debug-toolbar':
      'https://django-debug-toolbar.readthedocs.io/en/latest/changes.html',
    'firebase-admin':
      'https://firebase.google.com/support/release-notes/admin/python',
    requests:
      'http://docs.python-requests.org/en/master/community/updates/#release-and-version-history',
  },
};

// Use this object to define manual source URLs for packages
// Only necessary if the datasource is unable to locate the source URL itself
const manualSourceUrls = {
  orb: {
    'cypress-io/cypress': 'https://github.com/cypress-io/circleci-orb',
    'hutson/library-release-workflows':
      'https://github.com/hyper-expanse/library-release-workflows',
  },
  docker: {
    node: 'https://github.com/nodejs/node',
  },
  kubernetes: {
    node: 'https://github.com/nodejs/node',
  },
  npm: {
    node: 'https://github.com/nodejs/node',
  },
  nvm: {
    node: 'https://github.com/nodejs/node',
  },
  pypi: {
    coverage: 'https://github.com/nedbat/coveragepy/', // bitbucket entry on pypi is wrong
    mkdocs: 'https://github.com/mkdocs/mkdocs',
    pillow: 'https://github.com/python-pillow/Pillow',
  },
};

/* eslint-disable no-param-reassign */
function addMetaData(dep, datasource, lookupName) {
  if (!dep) {
    return;
  }
  const depName = lookupName ? lookupName.toLowerCase() : null;
  if (
    manualChangelogUrls[datasource] &&
    manualChangelogUrls[datasource][depName]
  ) {
    dep.changelogUrl = manualChangelogUrls[datasource][depName];
  }
  if (manualSourceUrls[datasource] && manualSourceUrls[datasource][depName]) {
    dep.sourceUrl = manualSourceUrls[datasource][depName];
  }
  if (
    !dep.sourceUrl &&
    dep.changelogUrl &&
    dep.changelogUrl.startsWith('https://github.com/')
  ) {
    dep.sourceUrl = dep.changelogUrl
      .split('/')
      .slice(0, 5)
      .join('/');
  }
  // Clean up any empty urls
  const urls = ['homepage', 'sourceUrl', 'changelogUrl'];
  for (const url of urls) {
    if (!is.nonEmptyString(dep[url])) {
      delete dep[url];
    }
  }
}
