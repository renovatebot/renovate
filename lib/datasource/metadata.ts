import is from '@sindresorhus/is';
import parse from 'github-url-from-git';
import { ReleaseResult } from './common';
import * as hostRules from '../util/host-rules';

// Use this object to define changelog URLs for packages
// Only necessary when the changelog data cannot be found in the package's source repository
const manualChangelogUrls = {
  npm: {
    'babel-preset-react-app':
      'https://github.com/facebook/create-react-app/releases',
    firebase: 'https://firebase.google.com/support/release-notes/js',
    'flow-bin': 'https://github.com/facebook/flow/blob/master/Changelog.md',
    gatsby:
      'https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/CHANGELOG.md',
    'react-native':
      'https://github.com/react-native-community/react-native-releases/blob/master/CHANGELOG.md',
  },
  pypi: {
    'pytest-django':
      'https://pytest-django.readthedocs.io/en/latest/changelog.html#changelog',
    django: 'https://github.com/django/django/tree/master/docs/releases',
    djangorestframework:
      'https://www.django-rest-framework.org/community/release-notes/',
    flake8: 'http://flake8.pycqa.org/en/latest/release-notes/index.html',
    'django-storages':
      'https://github.com/jschneier/django-storages/blob/master/CHANGELOG.rst',
    phonenumbers:
      'https://github.com/daviddrysdale/python-phonenumbers/blob/dev/python/HISTORY.md',
    'psycopg2-binary': 'http://initd.org/psycopg/articles/tag/release/',
    'django-debug-toolbar':
      'https://django-debug-toolbar.readthedocs.io/en/latest/changes.html',
    'firebase-admin':
      'https://firebase.google.com/support/release-notes/admin/python',
    requests:
      'http://docs.python-requests.org/en/master/community/updates/#release-and-version-history',
    wagtail: 'https://github.com/wagtail/wagtail/tree/master/docs/releases',
  },
};

// Use this object to define manual source URLs for packages
// Only necessary if the datasource is unable to locate the source URL itself
const manualSourceUrls = {
  orb: {
    'cypress-io/cypress': 'https://github.com/cypress-io/circleci-orb',
    'hutson/library-release-workflows':
      'https://github.com/hyper-expanse/library-release-workflows',
  },
  docker: {
    node: 'https://github.com/nodejs/node',
  },
  kubernetes: {
    node: 'https://github.com/nodejs/node',
  },
  npm: {
    node: 'https://github.com/nodejs/node',
  },
  nvm: {
    node: 'https://github.com/nodejs/node',
  },
  pypi: {
    coverage: 'https://github.com/nedbat/coveragepy/', // bitbucket entry on pypi is wrong
    mkdocs: 'https://github.com/mkdocs/mkdocs',
    pillow: 'https://github.com/python-pillow/Pillow',
  },
};

/* eslint-disable no-param-reassign */
export function addMetaData(
  dep?: ReleaseResult,
  datasource?: string,
  lookupName?: string
): void {
  if (!dep) {
    return;
  }
  const lookupNameLowercase = lookupName ? lookupName.toLowerCase() : null;
  if (
    manualChangelogUrls[datasource] &&
    manualChangelogUrls[datasource][lookupNameLowercase]
  ) {
    dep.changelogUrl = manualChangelogUrls[datasource][lookupNameLowercase];
  }
  if (
    manualSourceUrls[datasource] &&
    manualSourceUrls[datasource][lookupNameLowercase]
  ) {
    dep.sourceUrl = manualSourceUrls[datasource][lookupNameLowercase];
  }

  /**
   * @param {string} url
   */
  const massageGithubUrl = (url: string): string => {
    return url
      .replace('http:', 'https:')
      .replace(/^git:\/?\/?/, 'https://')
      .replace('www.github.com', 'github.com')
      .split('/')
      .slice(0, 5)
      .join('/');
  };
  if (
    dep.changelogUrl &&
    dep.changelogUrl.includes('github.com') && // lgtm [js/incomplete-url-substring-sanitization]
    !dep.sourceUrl
  ) {
    dep.sourceUrl = dep.changelogUrl;
  }
  // prettier-ignore
  if (dep.homepage && dep.homepage.includes('github.com')) { // lgtm [js/incomplete-url-substring-sanitization]
    if (!dep.sourceUrl) {
      dep.sourceUrl = dep.homepage;
    }
    delete dep.homepage;
  }
  const extraBaseUrls = [];
  // istanbul ignore next
  hostRules.hosts({ hostType: 'github' }).forEach(host => {
    extraBaseUrls.push(host, `gist.${host}`);
  });
  if (dep.sourceUrl) {
    // try massaging it
    dep.sourceUrl =
      parse(massageGithubUrl(dep.sourceUrl), {
        extraBaseUrls,
      }) || dep.sourceUrl;
  }

  // Clean up any empty urls
  const urls = ['homepage', 'sourceUrl', 'changelogUrl'];
  for (const url of urls) {
    if (is.nonEmptyString(dep[url])) {
      dep[url] = dep[url].trim();
      // istanbul ignore if
      if (!dep[url].match(/^https?:\/\//)) {
        delete dep[url];
      }
    } else {
      delete dep[url];
    }
  }
}
