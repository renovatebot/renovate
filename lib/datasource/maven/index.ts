import is from '@sindresorhus/is';
import url from 'url';
import fs from 'fs-extra';
import { XmlDocument } from 'xmldoc';
import { logger } from '../../logger';
import { compare } from '../../versioning/maven/compare';
import { containsPlaceholder } from '../../manager/maven/extract';
import { downloadHttpProtocol } from './util';
import { PkgReleaseConfig, ReleaseResult } from '../common';

export async function getPkgReleases({
  lookupName,
  registryUrls,
}: PkgReleaseConfig): Promise<ReleaseResult | null> {
  const versions: string[] = [];
  const dependency = getDependencyParts(lookupName);
  if (!is.nonEmptyArray(registryUrls)) {
    logger.warn(`No repositories defined for ${dependency.display}`);
    return null;
  }
  const repositories = registryUrls.map(repository =>
    repository.replace(/\/?$/, '/')
  );
  logger.debug(
    `Found ${repositories.length} repositories for ${dependency.display}`
  );
  const repoForVersions = {};
  for (let i = 0; i < repositories.length; i += 1) {
    const repoUrl = repositories[i];
    logger.debug(
      `Looking up ${dependency.display} in repository #${i} - ${repoUrl}`
    );
    const mavenMetadata = await downloadMavenXml(
      dependency,
      repoUrl,
      'maven-metadata.xml'
    );
    if (mavenMetadata) {
      const newVersions = extractVersions(mavenMetadata).filter(
        version => !versions.includes(version)
      );
      const latestVersion = getLatestVersion(newVersions);
      if (latestVersion) {
        repoForVersions[latestVersion] = repoUrl;
      }
      versions.push(...newVersions);
      logger.debug(`Found ${newVersions.length} new versions for ${dependency.display} in repository ${repoUrl}`); // prettier-ignore
    }
  }

  if (versions.length === 0) {
    logger.info(`No versions found for ${dependency.display} in ${repositories.length} repositories`); // prettier-ignore
    return null;
  }
  logger.debug(`Found ${versions.length} versions for ${dependency.display}`);
  const latestVersion = getLatestVersion(versions);
  const repoUrl = repoForVersions[latestVersion];
  const dependencyInfo = await getDependencyInfo(
    dependency,
    repoUrl,
    latestVersion
  );

  return {
    ...dependency,
    ...dependencyInfo,
    releases: versions.map(v => ({ version: v })),
  };
}

function getDependencyParts(lookupName: string): MavenDependency {
  const [group, name] = lookupName.split(':');
  const dependencyUrl = `${group.replace(/\./g, '/')}/${name}`;
  return {
    display: lookupName,
    group,
    name,
    dependencyUrl,
  };
}

interface MavenDependency {
  display: string;
  group?: string;
  name?: string;
  dependencyUrl: string;
}

async function downloadMavenXml(
  dependency: MavenDependency,
  repoUrl: string,
  dependencyFilePath: string
): Promise<XmlDocument | null> {
  let pkgUrl;
  try {
    pkgUrl = new url.URL(
      `${dependency.dependencyUrl}/${dependencyFilePath}`,
      repoUrl
    );
  } catch (err) {
    logger.debug(
      { err, dependency, repoUrl, dependencyFilePath },
      `Error constructing URL for ${dependency.display}`
    );
    return null;
  }

  let rawContent: string;
  switch (pkgUrl.protocol) {
    case 'file:':
      rawContent = await downloadFileProtocol(pkgUrl);
      break;
    case 'http:':
    case 'https:':
      rawContent = await downloadHttpProtocol(pkgUrl);
      break;
    case 's3:':
      logger.debug('Skipping s3 dependency');
      return null;
    default:
      logger.warn(
        `Invalid protocol ${pkgUrl.protocol} in repository ${repoUrl}`
      );
      return null;
  }

  if (!rawContent) {
    logger.debug(`${dependency.display} not found in repository ${repoUrl}`);
    return null;
  }

  try {
    return new XmlDocument(rawContent);
  } catch (e) {
    logger.debug(`Can not parse ${pkgUrl.href} for ${dependency.display}`);
    return null;
  }
}

function extractVersions(metadata: XmlDocument): string[] {
  const versions = metadata.descendantWithPath('versioning.versions');
  const elements = versions && versions.childrenNamed('version');
  if (!elements) return [];
  return elements.map(el => el.val);
}

async function downloadFileProtocol(pkgUrl: url.URL): Promise<string | null> {
  const pkgPath = pkgUrl.toString().replace('file://', '');
  if (!(await fs.exists(pkgPath))) {
    return null;
  }
  return fs.readFile(pkgPath, 'utf8');
}

function getLatestVersion(versions: string[]): string | null {
  if (versions.length === 0) return null;
  return versions.reduce((latestVersion, version) =>
    compare(version, latestVersion) === 1 ? version : latestVersion
  );
}

async function getDependencyInfo(
  dependency: MavenDependency,
  repoUrl: string,
  version: string
): Promise<Partial<ReleaseResult>> {
  const result: Partial<ReleaseResult> = {};
  const path = `${version}/${dependency.name}-${version}.pom`;

  const pomContent = await downloadMavenXml(dependency, repoUrl, path);
  if (!pomContent) return result;

  const homepage = pomContent.valueWithPath('url');
  if (homepage && !containsPlaceholder(homepage)) {
    result.homepage = homepage;
  }

  const sourceUrl = pomContent.valueWithPath('scm.url');
  if (sourceUrl && !containsPlaceholder(sourceUrl)) {
    result.sourceUrl = sourceUrl.replace(/^scm:/, '');
  }

  return result;
}
