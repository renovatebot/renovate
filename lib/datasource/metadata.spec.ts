import * as datasourceMaven from './maven';
import { addMetaData } from './metadata';
import * as datasourceNpm from './npm';
import * as datasourcePypi from './pypi';

describe('datasource/metadata', () => {
  it('Should do nothing if dep is not specified', () => {
    expect(addMetaData()).toBeUndefined();
  });

  it('Should handle manualChangelogUrls', () => {
    const dep = {
      releases: [
        { version: '2.0.0', releaseTimestamp: '2018-07-13T10:14:17.000Z' },
        {
          version: '2.0.0.dev1',
          releaseTimestamp: '2017-10-24T10:09:16.000Z',
        },
        { version: '2.1.0', releaseTimestamp: '2019-01-20T19:59:28.000Z' },
        { version: '2.2.0', releaseTimestamp: '2019-07-16T18:29:00.000Z' },
      ],
    };

    const datasource = datasourcePypi.id;
    const lookupName = 'django';

    addMetaData(dep, datasource, lookupName);
    expect(dep).toMatchSnapshot();
  });

  it('Should handle manualSourceUrls', () => {
    const dep = {
      releases: [
        { version: '2.0.0', releaseTimestamp: '2018-07-13T10:14:17.000Z' },
        {
          version: '2.0.0.dev1',
          releaseTimestamp: '2017-10-24T10:09:16.000Z',
        },
        { version: '2.1.0', releaseTimestamp: '2019-01-20T19:59:28.000Z' },
        { version: '2.2.0', releaseTimestamp: '2019-07-16T18:29:00.000Z' },
      ],
    };

    const datasource = datasourcePypi.id;
    const lookupName = 'mkdocs';

    addMetaData(dep, datasource, lookupName);
    expect(dep).toMatchSnapshot();
  });

  it('Should handle parsing of sourceUrls correctly', () => {
    const dep = {
      sourceUrl: 'https://github.com/carltongibson/django-filter/tree/master',
      releases: [
        { version: '2.0.0', releaseTimestamp: '2018-07-13T10:14:17.000Z' },
        {
          version: '2.0.0.dev1',
          releaseTimestamp: '2017-10-24T10:09:16.000Z',
        },
        { version: '2.1.0', releaseTimestamp: '2019-01-20T19:59:28.000Z' },
        { version: '2.2.0', releaseTimestamp: '2019-07-16T18:29:00.000Z' },
      ],
    };
    const datasource = datasourcePypi.id;
    const lookupName = 'django-filter';

    addMetaData(dep, datasource, lookupName);
    expect(dep).toMatchSnapshot();
  });

  it('Should handle parsing of sourceUrls correctly for GitLab also', () => {
    const dep = {
      sourceUrl: 'https://gitlab.com/meno/dropzone/tree/master',
      releases: [
        { version: '5.7.0', releaseTimestamp: '2020-02-14T13:12:00.000Z' },
        {
          version: '5.6.1',
          releaseTimestamp: '2020-02-14T10:04:00.000Z',
        },
      ],
    };
    const datasource = datasourceNpm.id;
    const lookupName = 'dropzone';

    addMetaData(dep, datasource, lookupName);
    expect(dep).toMatchSnapshot();
  });
  it('Should handle failed parsing of sourceUrls for GitLab', () => {
    const dep = {
      sourceUrl: 'https://gitlab-nope',
      releases: [
        { version: '5.7.0', releaseTimestamp: '2020-02-14T13:12:00.000Z' },
        {
          version: '5.6.1',
          releaseTimestamp: '2020-02-14T10:04:00.000Z',
        },
      ],
    };
    const datasource = datasourceNpm.id;
    const lookupName = 'dropzone';

    addMetaData(dep, datasource, lookupName);
    expect(dep).toMatchSnapshot();
  });
  it('Should handle failed parsing of sourceUrls for other', () => {
    const dep = {
      sourceUrl: 'https://nope-nope-nope',
      releases: [
        { version: '5.7.0', releaseTimestamp: '2020-02-14T13:12:00.000Z' },
        {
          version: '5.6.1',
          releaseTimestamp: '2020-02-14T10:04:00.000Z',
        },
      ],
    };
    const datasource = datasourceNpm.id;
    const lookupName = 'dropzone';

    addMetaData(dep, datasource, lookupName);
    expect(dep).toMatchSnapshot();
  });
  it('Should handle non-url', () => {
    const dep = {
      sourceUrl: 'not-a-url',
      releases: [
        { version: '5.7.0', releaseTimestamp: '2020-02-14T13:12:00.000Z' },
        {
          version: '5.6.1',
          releaseTimestamp: '2020-02-14T10:04:00.000Z',
        },
      ],
    };
    const datasource = datasourceNpm.id;
    const lookupName = 'dropzone';

    addMetaData(dep, datasource, lookupName);
    expect(dep).toMatchSnapshot();
  });

  it('Should handle parsing/converting of GitHub sourceUrls with http and www correctly', () => {
    const dep = {
      sourceUrl: 'http://www.github.com/mockk/mockk/',
      releases: [{ version: '1.9.3' }],
    };
    const datasource = datasourceMaven.id;
    const lookupName = 'io.mockk:mockk';

    addMetaData(dep, datasource, lookupName);
    expect(dep.sourceUrl).toEqual('https://github.com/mockk/mockk');
  });

  it('Should move github homepage to sourceUrl', () => {
    const dep = {
      homepage: 'http://www.github.com/mockk/mockk/',
      releases: [{ version: '1.9.3' }],
      sourceUrl: undefined,
    };
    const datasource = datasourceMaven.id;
    const lookupName = 'io.mockk:mockk';

    addMetaData(dep, datasource, lookupName);
    expect(dep.sourceUrl).toEqual('https://github.com/mockk/mockk');
    expect(dep.homepage).toBeUndefined();
  });

  it('Should handle parsing/converting of GitLab sourceUrls with http and www correctly', () => {
    const dep = {
      sourceUrl: 'http://gitlab.com/meno/dropzone/',
      releases: [{ version: '5.7.0' }],
    };
    const datasource = datasourceMaven.id;
    const lookupName = 'dropzone';

    addMetaData(dep, datasource, lookupName);
    expect(dep.sourceUrl).toEqual('https://gitlab.com/meno/dropzone');
  });

  it('Should normalize releaseTimestamp', () => {
    const dep = {
      releases: [
        { version: '1.0.1', releaseTimestamp: '2000-01-01T12:34:56' },
        { version: '1.0.2', releaseTimestamp: '2000-01-02T12:34:56.000Z' },
        { version: '1.0.3', releaseTimestamp: '2000-01-03T14:34:56.000+02:00' },
      ],
    };
    addMetaData(dep, datasourceMaven.id, 'foobar');
    expect(dep.releases).toMatchObject([
      { releaseTimestamp: '2000-01-01T12:34:56.000Z' },
      { releaseTimestamp: '2000-01-02T12:34:56.000Z' },
      { releaseTimestamp: '2000-01-03T12:34:56.000Z' },
    ]);
  });
});
