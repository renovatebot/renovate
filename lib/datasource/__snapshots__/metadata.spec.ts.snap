// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datasource/metadata Should handle failed parsing of sourceUrls for GitLab 1`] = `
Object {
  "releases": Array [
    Object {
      "releaseTimestamp": "2020-02-14T13:12:00.000Z",
      "version": "5.7.0",
    },
    Object {
      "releaseTimestamp": "2020-02-14T10:04:00.000Z",
      "version": "5.6.1",
    },
  ],
  "sourceUrl": "https://gitlab-nope",
}
`;

exports[`datasource/metadata Should handle failed parsing of sourceUrls for other 1`] = `
Object {
  "releases": Array [
    Object {
      "releaseTimestamp": "2020-02-14T13:12:00.000Z",
      "version": "5.7.0",
    },
    Object {
      "releaseTimestamp": "2020-02-14T10:04:00.000Z",
      "version": "5.6.1",
    },
  ],
  "sourceUrl": "https://nope-nope-nope",
}
`;

exports[`datasource/metadata Should handle manualChangelogUrls 1`] = `
Object {
  "changelogUrl": "https://github.com/django/django/tree/master/docs/releases",
  "releases": Array [
    Object {
      "releaseTimestamp": "2018-07-13T10:14:17.000Z",
      "version": "2.0.0",
    },
    Object {
      "releaseTimestamp": "2017-10-24T10:09:16.000Z",
      "version": "2.0.0.dev1",
    },
    Object {
      "releaseTimestamp": "2019-01-20T19:59:28.000Z",
      "version": "2.1.0",
    },
    Object {
      "releaseTimestamp": "2019-07-16T18:29:00.000Z",
      "version": "2.2.0",
    },
  ],
  "sourceUrl": "https://github.com/django/django",
}
`;

exports[`datasource/metadata Should handle manualSourceUrls 1`] = `
Object {
  "releases": Array [
    Object {
      "releaseTimestamp": "2018-07-13T10:14:17.000Z",
      "version": "2.0.0",
    },
    Object {
      "releaseTimestamp": "2017-10-24T10:09:16.000Z",
      "version": "2.0.0.dev1",
    },
    Object {
      "releaseTimestamp": "2019-01-20T19:59:28.000Z",
      "version": "2.1.0",
    },
    Object {
      "releaseTimestamp": "2019-07-16T18:29:00.000Z",
      "version": "2.2.0",
    },
  ],
  "sourceUrl": "https://github.com/mkdocs/mkdocs",
}
`;

exports[`datasource/metadata Should handle non-url 1`] = `
Object {
  "releases": Array [
    Object {
      "releaseTimestamp": "2020-02-14T13:12:00.000Z",
      "version": "5.7.0",
    },
    Object {
      "releaseTimestamp": "2020-02-14T10:04:00.000Z",
      "version": "5.6.1",
    },
  ],
}
`;

exports[`datasource/metadata Should handle parsing of sourceUrls correctly 1`] = `
Object {
  "releases": Array [
    Object {
      "releaseTimestamp": "2018-07-13T10:14:17.000Z",
      "version": "2.0.0",
    },
    Object {
      "releaseTimestamp": "2017-10-24T10:09:16.000Z",
      "version": "2.0.0.dev1",
    },
    Object {
      "releaseTimestamp": "2019-01-20T19:59:28.000Z",
      "version": "2.1.0",
    },
    Object {
      "releaseTimestamp": "2019-07-16T18:29:00.000Z",
      "version": "2.2.0",
    },
  ],
  "sourceUrl": "https://github.com/carltongibson/django-filter",
}
`;

exports[`datasource/metadata Should handle parsing of sourceUrls correctly for GitLab also 1`] = `
Object {
  "releases": Array [
    Object {
      "releaseTimestamp": "2020-02-14T13:12:00.000Z",
      "version": "5.7.0",
    },
    Object {
      "releaseTimestamp": "2020-02-14T10:04:00.000Z",
      "version": "5.6.1",
    },
  ],
  "sourceUrl": "https://gitlab.com/meno/dropzone",
}
`;
